infrastructure/smart-contracts/contracts/upgrades/pool-factory/PoolFactory3.sol

```diff
@@ -0,0 +1,875 @@
+// 34: Invalid delegate update; pool needs to have been created via the PoolFactory
+// 35: Invalid submission; pool needs to have been created via the PoolFactory
+// 36: Invalid submission; user needs to be owner, delegate or staker to submit
+// 37: Invalid auth: msg.sender must be kyc approved to create a pool
+// 38: Address failed kyc
+
+contract PoolFactory3 is Initializable, AccessControlEnumerableUpgradeable {
+    using EnumerableSetUpgradeable for EnumerableSetUpgradeable.AddressSet;
+
+    // address of the NodeLicense NFT contract
+    address public nodeLicenseAddress;
+
+    // contract addresses for esXai and xai tokens
+    address public esXaiAddress;
+    address public refereeAddress;
+
+    // Boolean to enable/disable staking on the referee contract
+    bool public stakingEnabled;
+
+    // Array to store addresses of staking pool contracts
+    address[] public stakingPools;
+
+    // Maximum share values for staking pool (owner, keys, stakedEsXai) in basepoints (5% => 50,000)
+    uint32[3] public bucketshareMaxValues;
+
+    // Mapping of user addresses to their interacted pool addresses
+    mapping(address => address[]) public interactedPoolsOfUser;
+
+    // Mapping of user addresses to pool addresses to index in user array for efficient removal
+    mapping(address => mapping(address => uint256))
+        public userToInteractedPoolIds;
+
+    // Mapping of delegates to the pools they are delegates of
+    mapping(address => address[]) public poolsOfDelegate;
+
+    // Mapping of pool addresses to indices in the delegate's pools array
+    mapping(address => uint256) public poolsOfDelegateIndices;
+
+    // Mapping of pool addresses created via this factory
+    mapping(address => bool) public poolsCreatedViaFactory;
+
+    // Address of the contract responsible for deploying staking pool and bucket proxies
+    address public deployerAddress;
+
+    // Periods (in seconds) to lock keys/esXai when user creates an unstake request
+    uint256 public unstakeKeysDelayPeriod;
+    uint256 public unstakeGenesisKeyDelayPeriod;
+    uint256 public unstakeEsXaiDelayPeriod;
+
+    // Period (in seconds) to update reward breakdown changes
+    uint256 public updateRewardBreakdownDelayPeriod;
+
+    // Mapping to track if an address failed kyc
+    mapping(address => bool) public failedKyc; // Default is false
+
+    // Role definition for stake keys admin
+    bytes32 public constant STAKE_KEYS_ADMIN_ROLE = keccak256("STAKE_KEYS_ADMIN_ROLE");
+
+    mapping(address => bool) public totalStakeCalculated;
+
+    // =================> VERY IMPORTANT <============================================
+    // Making this variable private as it SHOULD NOT BE USED as the source of truth for total stake
+    // The getTotalesXaiStakedByUser function should be used instead
+    // This mapping will only be accurate AFTER a user has interacted with a pool
+    mapping(address => uint256) private _totalStakeByUser;
```

Comment:
This should have esXai in the name, users also stake keys

Reply:
Done!